### ISSUE TITLE
faq gives incorrect instructions for running typeorm CLI with ts-node

### ISSUE THREAD

**Issue by binki** [typeorm/typeorm#2316]
**Issue type:**

[ ] question
[ ] bug report
[ ] feature request
[x] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `0.2.7` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I was reading https://github.com/typeorm/typeorm/blob/966b88665f3cb1a0796ea42296980312febe80d5/docs/faq.md#how-to-use-typeorm-with-ts-node and found two issues.

1. The path into `node_modules/.bin` was mistyped as `node_modules/bin`. `npm` installs binaries to `node_modules/.bin`. This is probably done to prevent a possible collision with a package name. Starting the folder with a `.` is safe because [`package.json#name`](https://docs.npmjs.com/files/package.json#name) mayn’t start with a period or underscore. And it is totally possible that a [package literally named `bin`](https://www.npmjs.com/package/bin) would exist.
2. Even if using the corrected path `node_modules/.bin/typeorm`, that path is not a JavaScript file on Windows. It is a shell script to be run when using MSYS2/cywgin `bash` on Windows with an adjacent `.cmd` which is used by `CMD` if in `CMD`.

```
mkdir my-repo
cd my-repo
npm init -y
npm install -S typeorm ts-node typescript
```

I prefer to avoid installing npm packages globally. So that is why I installed ts-node to my repo instead of globally. I understand that I must manually type the path to it because it won’t be in my PATH. Thus, I am voluntarily choosing to type the path to `ts-node` and that is not related to the issue I am demonstrating.

In MSYS2 `bash`, I can type the command as if I were on POSIX:

```
$ node_modules/.bin/ts-node node_modules/.bin/typeorm --help
C:\Users\ohnob\AppData\Local\Temp\my-repo\node_modules\.bin\typeorm:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:616:28)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Function.Module.runMain (module.js:693:10)
    at Object.<anonymous> (C:\Users\ohnob\AppData\Local\Temp\my-repo\node_modules\ts-node\src\bin.ts:147:12)
    at Module._compile (module.js:652:30)
$ echo $?
1
```

In CMD, I type the `ts-node` command using backslashes because `CMD` considers the first forwardslash to be a parameter even without whitespace. Again, that is besides the point of it being impossible to invoke the `typeorm` CLI via `ts-node` this way:

```
C:\Users\ohnob\AppData\Local\Temp\my-repo>node_modules\.bin\ts-node node_modules/.bin/typeorm
C:\Users\ohnob\AppData\Local\Temp\my-repo\node_modules\.bin\typeorm:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:616:28)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Function.Module.runMain (module.js:693:10)
    at Object.<anonymous> (C:\Users\ohnob\AppData\Local\Temp\my-repo\node_modules\ts-node\src\bin.ts:147:12)
    at Module._compile (module.js:652:30)

C:\Users\ohnob\AppData\Local\Temp\my-repo>ECHO %ERRORLEVEL%
1

```

So, the issue is that `ts-node` needs to be passed a path to a JavaScript file instead of an executable which may or may not be JavaScript depending on the platform.

My understanding is that `ts-node` or something similar is required because the `typeorm` CLI will need to load your untranspiled `migrations/*.ts` or `entities/*.ts` to run certain commands. If running as vanilla node, it can’t know to do this.

There are two approaches to doing this successfully with support for Windows.

1. Always place the `cli.js` in the same location in the package and recommend users run `ts-node node_modules/typeorm/cli.js` instead of `ts-node node_modules/.bin/typeorm` in the FAQ.
2. Add something like the `-r` option to `typeorm`’s CLI. So then you can instruct the user to run `node_modules/.bin/typeorm -r ts-node/register` in the FAQ. I don’t think the CLI has such a feature yet but, in a way, this would both avoid vendor locking to `ts-node` versus some other package offering the same feature and avoid needing to always keep the path of `cli.js` the same within the published `typeorm` package.

**Comment by Slessi** [comment]
Personally I opt for `ts-node $(npm bin)/typeorm`

**Comment by binki** [comment]
@Slessi That might be a bit harder for Windows users to translate into something `CMD` would understand. For example, they could do:

```
>npm bin>.my_npm_bin
>SET /P MY_NPM_BIN=<.my_npm_bin~
>DEL /Q .my_npm_bin~
>%MY_NPM_BIN%\ts-node %MY_NPM_BIN%\typeorm
```

And then you still get this output:

```
C:\Users\ohnob\repos\«redacted»\node_modules\.bin\typeorm:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:616:28)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Function.Module.runMain (module.js:693:10)
    at Object.<anonymous> (C:\Users\ohnob\repos\«redacted»\node_modules\ts-node\src\bin.ts:147:12)
    at Module._compile (module.js:652:30)
```

**Comment by KishanIos007** [comment]
I am having this issue in one my ionic 1 project I am using yarn package manager .. and windows 10 .. can anyone help me out ?

**Comment by thismarcoantonio** [comment]
Any updates on this?

**Comment by pleerock** [comment]
@hkjeffchan @wederer since you guys were updating documentation can you please take a look on this issue as well?

**Comment by thismarcoantonio** [comment]
I figured out what was the error. In windows machines, using `./node_modules/.bin/typeorm` doesn't work, so I used the cli.js found in typeorm folder, in node_modules. I think it is an alternative to typeorm, written in js.

**Comment by acidduckling** [comment]
> I figured out what was the error. In windows machines, using `./node_modules/.bin/typeorm` doesn't work, so I used the cli.js found in typeorm folder, in node_modules. I think it is an alternative to typeorm, written in js.

This seems to have worked for me. Thanks @thismarcoantonio

**Comment by hkjeffchan** [comment]
use cli.js is better, it should work on both mac and windows (although i don't have a pc to verify this)

update doc with  #3433

**Comment by thismarcoantonio** [comment]
@hkjeffchan there is any reason to not deprecate the `./node_modules/.bin/typeorm` and use only `cli.js`?

**Comment by pleerock** [comment]
fixed by #3433
